File: 3619-adjacent-increasing-subarrays-detection-ii/adjacent-increasing-subarrays-detection-ii.cpp
#define ll long long
#define vi vector<int>
#define vll vector<ll>
#define vs vector<string>
#define mpi map<int, int>
#define mps map<string, int>
#define pb push_back
#define all(x) (x).begin(), (x).end()

const int MOD = 1e9 + 7;

class Solution {
public:
    int maxIncreasingSubarrays(vector<int>& nums) {
        int n = nums.size();
        int up = 1, preUp = 0;
        int ans = 0;

        for (int i = 1; i < n; i++) {
            if (nums[i] > nums[i - 1]) {
                up++;
            } else {
                preUp = up;
                up = 1;
            }

            int half = up >> 1;
            int m = min(preUp, up);
            int maxc = max(half, m);
            ans = max(ans, maxc);
        }

        return ans;
    }
};

File: 3619-adjacent-increasing-subarrays-detection-ii/adjacent-increasing-subarrays-detection-ii.py
class Solution:
    def maxIncreasingSubarrays(self, nums: list[int]) -> int:
        n = len(nums)
        inc, preinc, ans = 1, 0, 0

        for i in range(1, n):
            if nums[i - 1] < nums[i]:
                inc += 1
            else:
                preinc = inc
                inc = 1

            half = inc >> 1
            mn = min(preinc, inc)
            maxc = max(half, mn)
            ans = max(ans, maxc)

        return ans
